src/compiler/checker.ts

```diff
@@ -16535,6 +16517,89 @@ namespace ts {
+        // If this is a type resolution session, e.g. Language Service, try to get better information than anySignature.
+        function getCandidateForOverloadFailure(
+            node: CallLikeExpression,
+            candidates: Signature[],
+            args: ReadonlyArray<Expression>,
+            hasCandidatesOutArray: boolean,
+        ): Signature {
+            Debug.assert(candidates.length > 0); // Else should not have called this.
+
+            // Normally we will combine overloads. Skip this if they have type parameters since that's hard to combine.
+            // Don't do this if there is a `candidatesOutArray`,
+            // because then we want the chosen best candidate to be one of the overloads, not a combination.
+            if (!hasCandidatesOutArray && candidates.length > 1 && !candidates.some(c => !!c.typeParameters)) {
+                return createUnionOfSignaturesForOverloadFailure(candidates);
+            }
+
+            // Pick the longest signature. This way we can get a contextual type for cases like:
+            //     declare function f(a: { xa: number; xb: number; }, b: number);
+            //     f({ |
+            // Also, use explicitly-supplied type arguments if they are provided, so we can get a contextual signature in cases like:
+            //     declare function f<T>(k: keyof T);
+            //     f<Foo>("
+            const bestIndex = getLongestCandidateIndex(candidates, apparentArgumentCount === undefined ? args.length : apparentArgumentCount);
+            const candidate = candidates[bestIndex];
+
+            const { typeParameters } = candidate;
+            if (typeParameters && callLikeExpressionMayHaveTypeArguments(node) && node.typeArguments) {
+                const typeArguments = node.typeArguments.map(getTypeOfNode);
+                while (typeArguments.length > typeParameters.length) {
+                    typeArguments.pop();
+                }
+                while (typeArguments.length < typeParameters.length) {
+                    typeArguments.push(getDefaultTypeArgumentType(isInJavaScriptFile(node)));
+                }
+
+                const instantiated = createSignatureInstantiation(candidate, typeArguments);
+                candidates[bestIndex] = instantiated;
+                return instantiated;
+            }
+
+            return candidate;
+        }
+
+        function createUnionOfSignaturesForOverloadFailure(candidates: ReadonlyArray<Signature>): Signature {
+            const thisParameters = mapDefined(candidates, c => c.thisParameter);
+            let thisParameter: Symbol | undefined;
+            if (thisParameters.length) {
+                thisParameter = createCombinedSymbolWithType(thisParameters, getUnionType(thisParameters.map(getTypeOfParameter), /*subtypeReduction*/ true));
+            }
+
+            const { min: minArgumentCount, max: maxNonRestParam }  = minAndMax(candidates, getNumNonRestParameters);

```

Comment:
linter should catch the double space? I think?

Reply:
There's no such rule currently. Note that we do equals-sign alignment in a few places so we'd have to change our style to conform to such a rule -- it might be worth it though.